File: javascript/src/problems101_200/maximum-depth-of-binary-tree.js
class Problem {
  maxDepth(root) {
    throw new Error(`Not implemented (${root}).`);
  }
}

class Default extends Problem {
  maxDepth(root) {
    return this.recursiveMaxDepth(root, 0);
  }

  recursiveMaxDepth(node, depth) {
    if (!node) {
      return depth;
    }
    const leftDepth = this.recursiveMaxDepth(node.left, depth + 1);
    const rightDepth = this.recursiveMaxDepth(node.right, depth + 1);
    return Math.max(leftDepth, rightDepth);
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems101_200/maximum_depth_of_binary_tree.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.tree_node import TreeNode


class Problem(ABC):
    @abstractmethod
    def max_depth(self, root: TreeNode | None) -> int:
        pass


class Default(Problem):
    @override
    def max_depth(self, root: TreeNode | None) -> int:
        return self.recursive_max_depth(root, 0)

    def recursive_max_depth(self, node: TreeNode | None, depth: int) -> int:
        if not node:
            return depth
        left_depth: int = self.recursive_max_depth(node.left, depth + 1)
        right_depth: int = self.recursive_max_depth(node.right, depth + 1)
        return max(left_depth, right_depth)


SOLUTIONS: list[Problem] = [Default()]
